pkg/crawl/crawler.go

```diff
@@ -60,55 +59,87 @@ type ProcessFunc func(ctx context.Context, path string, result Finding, version
 // the directory) will be ignored.
 func (c Crawler) Crawl(ctx context.Context, root string, match MatchFunc, process ProcessFunc) (Stats, error) {
 	stats := Stats{}
-	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
-		if err != nil {
-			switch {
-			case os.IsPermission(err):
-				stats.PermissionDeniedCount++
-				return nil
-			case os.IsNotExist(err):
-				// Root should always exist, to pick up on misconfigured service, but log4j-sniffer can encounter transient
-				// files when walking, where WalkDir lists the directory entries but the entry disappears before or
-				// during the walk function iterating over it.
-				if path == root {
-					return err
-				}
-				return nil
-			}
-			return err
-		}
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		default:
+	rootFile, err := os.Lstat(root)
+	if err != nil {
+		return stats, err
+	}
+	if !rootFile.IsDir() {
+		err = c.processFile(ctx, &stats, root, rootFile.Name(), match, process)
+	} else {
+		err = c.processDir(ctx, &stats, root, match, process)
+	}
+	return stats, err
+}
+
+func (c Crawler) processDir(ctx context.Context, stats *Stats, path string, match MatchFunc, process ProcessFunc) error {
+	dirInfo, err := os.Open(path)
+	switch {
+	case os.IsPermission(err):
+		stats.PermissionDeniedCount++
+		return nil
+	case os.IsNotExist(err):
+		return nil
+	case err != nil:
+		stats.PathErrorCount++
+		if c.ErrorWriter != nil {
+			_, _ = fmt.Fprintf(c.ErrorWriter, "Error processing path %s: %v\n", path, err)
 		}
-		if d.IsDir() {
-			if c.includeDir(path) {
+		return err
+	}
+	dirEntries, err := dirInfo.ReadDir(100)
+	for err != io.EOF {
+		for _, entry := range dirEntries {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+			}
+
+			nestedPath := strings.Join([]string{path, entry.Name()}, string(os.PathSeparator))
+			if entry.IsDir() {
+				if !c.includeDir(nestedPath) {
+					continue
+				}
 				c.Limiter.Take()
-				return nil
+				nestedError := c.processDir(ctx, stats, nestedPath, match, process)
+				if nestedError != nil {
+					return nestedError
+				}
+			} else if !entry.Type().IsRegular() {
+				continue
+			} else {
+				err = c.processFile(ctx, stats, nestedPath, entry.Name(), match, process)
+				if err != nil {
```

Comment:
Let's inline this and assign new `err` given that there is a lingering `err` in the for loop statement and reducing this scope will make the behaviour clearer.

```suggestion
				if err := c.processFile(ctx, stats, nestedPath, entry.Name(), match, process); err != nil {
```